src/common/hist_util.cu

```diff
@@ -341,63 +357,91 @@ struct GPUSketcher {
+                                   thrust::maximum<size_t>());
     }
   };
 
-  void Sketch(const SparsePage& batch, const MetaInfo& info,
-              HistCutMatrix* hmat, int gpu_batch_nrows) {
+  size_t SketchBatch(const GPUDistribution &dist, const SparsePage &batch,
+                     const MetaInfo &info, SketchContainer *sketch_container) {
     // create device shards
-    shards_.resize(dist_.Devices().Size());
+    shards_.resize(dist.Devices().Size());
     dh::ExecuteIndexShards(&shards_, [&](int i, std::unique_ptr<DeviceShard>& shard) {
-        size_t start = dist_.ShardStart(info.num_row_, i);
-        size_t size = dist_.ShardSize(info.num_row_, i);
+        size_t start = dist.ShardStart(batch.Size(), i);
+        size_t size = dist.ShardSize(batch.Size(), i);
         shard = std::unique_ptr<DeviceShard>(
-            new DeviceShard(dist_.Devices().DeviceId(i),
-                            start, start + size, param_));
+            new DeviceShard(dist.Devices().DeviceId(i), start,
+                            start + size, param_, sketch_container));
       });
 
     // compute sketches for each shard
     dh::ExecuteIndexShards(&shards_,
                            [&](int idx, std::unique_ptr<DeviceShard>& shard) {
-                             shard->Init(batch, info, gpu_batch_nrows);
+                             shard->Init(batch, info, gpu_batch_nrows_);
                              shard->Sketch(batch, info);
                            });
-
-    // merge the sketches from all shards
-    // TODO(canonizer): do it in a tree-like reduction
-    int num_cols = info.num_col_;
-    std::vector<WXQSketch> sketches(num_cols);
-    WXQSketch::SummaryContainer summary;
-    for (int icol = 0; icol < num_cols; ++icol) {
-      sketches[icol].Init(batch.Size(), 1.0 / (8 * param_.max_bin));
-      for (auto &shard : shards_) {
-        shard->GetSummary(&summary, icol);
-        sketches[icol].PushSummary(summary);
-      }
+    size_t row_stride = 0;
+    for (const auto &shard : shards_) {
+      row_stride = std::max(row_stride, shard->GetRowStride());
     }
 
-    hmat->Init(&sketches, param_.max_bin);
+    return row_stride;
+  }
+
+  GPUSketcher(const tree::TrainParam &param, int gpu_nrows)
+    : param_(param), gpu_batch_nrows_(gpu_nrows) {
   }
 
-  GPUSketcher(tree::TrainParam param, size_t n_rows) : param_(std::move(param)) {
-    dist_ = GPUDistribution::Block(GPUSet::All(param_.gpu_id, param_.n_gpus, n_rows));
+  /* Builds the sketches on the GPU */
+  size_t Sketch(DMatrix *dmat, HistCutMatrix *hmat) {
+    size_t row_stride = 0;
+    const MetaInfo &info = dmat->Info();
+
+    // Initialize Sketches for this dmatrix
+    SketchContainer sketch_container;
+    sketch_container.sketches_.resize(info.num_col_);
+    sketch_container.col_locks_.resize(info.num_col_);
+#pragma omp parallel for schedule(static) if (info.num_col_ > 1000)
+    for (int icol = 0; icol < info.num_col_; ++icol) {
```

Comment:
This looks like a constructor or initializer for SketchContainer.

Reply:
will do.